lib/index.js

```diff
@@ -2,23 +2,158 @@
  * Copyright (c) 2020 Digital Bazaar, Inc. All rights reserved.
  */
 'use strict';
-// const {generateId} = require('bnid');
-// const {snakeCase} = require('snake-case');
-// const {
-//   InvalidRequest, AccessDenied
-// } = require('@interop/oauth2-errors');
 
-function handleTokenRequest({logger}) {
+const {
+  UnauthorizedClient, InvalidGrant, InvalidRequest, InvalidClient
+} = require('@interop/oauth2-errors');
+
+// Default Access token expiration in seconds (two weeks)
+const DEFAULT_MAX_AGE = 1209600;
+
+/**
+ * Constructs a middleware handler for OAuth 2.0 /token exchange requests.
+ *
+ * @param {object} options - Options hashmap.
+ * @param {object} [options.defaults={}] - Token issuance defaults.
+ * @param {object} options.logger - Logger.
+ * @param {Function} options.getClient - Callback to load client by id.
+ * @param {Function} options.issue - Callback to issue the JWT access token for
+ *   client.
+ *
+ * @example <caption>Example usage.</caption>
+ * tokenExchangeHandler({
+ *   defaults: {},
+ *   logger: console,
+ *   getClient: async ({clientId}) => {
+ *     // load client from the db, then:
+ *     return {client};
+ *   }),
+ *   issue: async ({client, scope, body, defaultMaxAge}) => {
+ *     // issue JWT access token here
+ *     return {accessToken, expiresIn};
+ *   })
+  },
+ * });
+ *
+ * @returns {Promise<Function>} Resolves with exchange middleware.
+ */
+function tokenExchangeHandler({defaults = {}, logger, getClient, issue}) {
   return async (req, res) => {
     try {
-    } catch(error) {
-      if(logger) {
-        logger.error(error);
+      const {body} = req;
+      const {
+        client_id: clientId,
+        client_secret: clientSecret,
+        grant_type: grantType,
+        scope
+      } = body;
+
+      _validateParams({clientId, clientSecret, grantType, scope});
+
+      let client;
+      try {
+        client = await getClient({clientId});
+      } catch(cause) {
+        const error = new InvalidRequest({
+          description: 'Could not load client.'
+        });
+        error.cause = cause;
+        throw error;
       }
-      error.respond(res);
+
+      _authenticateClient({client, clientSecret});
```

Comment:
If we make this a callback/option to be passed in, we can push the secret comparison handling to the application (which we should probably do anyway, since it needs to be customizable).